crates/goose/src/agents/subagent_client.rs

```diff
@@ -0,0 +1,207 @@
+use crate::agents::tool_execution::ToolCallResult;
+use crate::config::get_enabled_extensions;
+use anyhow::Result;
+use async_trait::async_trait;
+use rmcp::model::{
+    CallToolResult, Content, GetPromptResult, Implementation, InitializeResult, JsonObject,
+    ListPromptsResult, ListResourcesResult, ListToolsResult, ProtocolVersion, ReadResourceResult,
+    ServerCapabilities, ServerNotification, Tool, ToolsCapability,
+};
+use serde_json::Value;
+use tokio::sync::mpsc;
+use tokio_util::sync::CancellationToken;
+
+pub const EXTENSION_NAME: &str = "subagent";
+
+pub struct SubagentClient {
+    context: PlatformExtensionContext,
+    info: InitializeResult,
+}
+
+impl SubagentClient {
+    pub fn new(context: PlatformExtensionContext) -> Result<Self> {
+        Ok(Self {
+            context,
+            info: InitializeResult {
+                protocol_version: ProtocolVersion::V_2025_03_26,
+                capabilities: ServerCapabilities {
+                    tools: Some(ToolsCapability {
+                        list_changed: Some(false),
+                    }),
+                    resources: None,
+                    prompts: None,
+                    completions: None,
+                    experimental: None,
+                    logging: None,
+                },
+                server_info: Implementation {
+                    name: EXTENSION_NAME.to_string(),
+                    title: Some("Subagent".to_string()),
+                    version: env!("CARGO_PKG_VERSION").to_string(),
+                    icons: None,
+                    website_url: None,
+                },
+                instructions: Some(
+                    "Delegate tasks to independent subagents for parallel or focused work."
+                        .to_string(),
+                ),
+            },
+        })
+    }
+
+    async fn get_provider(&self) -> Option<std::sync::Arc<dyn crate::providers::base::Provider>> {
+        let em = self.context.extension_manager.as_ref()?.upgrade()?;
+        em.get_provider().await
+    }
+
+    async fn get_extensions(&self) -> Vec<crate::agents::ExtensionConfig> {
+        if let Some(em) = self
+            .context
+            .extension_manager
+            .as_ref()
+            .and_then(|w| w.upgrade())
+        {
+            em.get_extension_configs().await
+        } else {
+            get_enabled_extensions()
+        }
+    }
+
+    async fn get_sub_recipes(&self) -> std::collections::HashMap<String, crate::recipe::SubRecipe> {
+        match &self.context.sub_recipes {
+            Some(recipes) => recipes.read().await.clone(),
+            None => std::collections::HashMap::new(),
+        }
+    }
+
+    async fn build_tool(&self) -> Tool {
+        let sub_recipes = self.get_sub_recipes().await;
+        let sub_recipes_vec: Vec<_> = sub_recipes.values().cloned().collect();
+        create_subagent_tool(&sub_recipes_vec)
+    }
+}
+
+#[async_trait]
+impl McpClientTrait for SubagentClient {
```

Comment:
just a thought (not feedback on this PR)

We could probably benefit from an extension construct that isn't necessarily an MCP extension. We're making good use of the platform extension construct, but having them "pretend" to be MCP servers results in us needing some odd code like this.